source/common/router/upstream_request.cc

```diff
@@ -223,6 +224,26 @@ const Network::Connection& UpstreamRequest::connection() const {
   return *parent_.callbacks()->connection();
 }
 
+void UpstreamRequest::copyDynamicMetaDataFromL4Downstream(StreamInfo::StreamInfo& l4_stream_info) {
+  if (parent_.callbacks()->connection() != nullptr) {
+    const StreamInfo::StreamInfo& downstream_streamInfo =
+        parent_.callbacks()->connection()->streamInfo();
+    const auto& dynamic_metadata = downstream_streamInfo.dynamicMetadata();
+    auto name = Envoy::Extensions::NetworkFilters::NetworkFilterNames::get().SharedWithUpstream;
+    auto filter_it = dynamic_metadata.filter_metadata().find(name);
+
+    if (filter_it != dynamic_metadata.filter_metadata().end()) {
+      auto downstream_connection_id = parent_.callbacks()->connection()->id();
+      ProtobufWkt::Struct upstream_metadata1;
+      auto& fields = *upstream_metadata1.mutable_fields();
+      ProtobufWkt::Value val;
+      val = ValueUtil::structValue(filter_it->second);
+      fields.insert({std::to_string(downstream_connection_id), val});
```

Comment:
Does this leak memory when downstream creates many connections through a single upstream connection? Is the upstream filter expected to delete a field?

Reply:
I will consider more. Thanks for the input.